### PR TITLE
[WIP] implement attention function and its tests

### PR DESCRIPTION
**Pull request by ywatanabex**
Hi, I have implemented an `Attention` function.   This PR is related to #1347.

This function computes the attention from vector to (variable length) sequence.
The output of the function is the average of feature vectors of the sequence w.r.t. attention probability.

I think the code has a lot of room for improvements. I'm very happy to improve it given reviews and feedbacks.

### CODE REVIEW

**Comment by unnonouno** [comment]
Thank you for your PR! I feel we need to discuss about interface and functionality of attention link.
- Is it possible to apply the layer to memory networks? In memory networks, variables to calculate weights and variables to merge are different (A and C in the paper 1).
- Luong et.al. compared different strategies to calculates weights eq. (8) in the paper 2. Do you have an idea to support arbitrary strategies in the link?

1. https://arxiv.org/abs/1503.08895
2. http://www.aclweb.org/anthology/D15-1166

**Comment by ywatanabex** [comment]
Hi. Thank you for your comments.
Regarding the point 1, I agree that it is a good generalization.
Maybe we can write

```python
>>> attention(q, as)             # if as = cs   (current interface)
>>> attention(q, as, cs)       # generalized version as in E2EMemNet
```
where as and cs are lists of sequences (chainer variables)

**Comment by ywatanabex** [comment]
Regarding the second point, I don't know how to support any function.

But for those three cases, we can think of ht^T Wa hs + w1^T ht + w2^T hu   
This form includes all three cases.
(dot: Wa=I, w1=w2=0,  concat: Wa=0)

**Comment by unnonouno** [comment]
The first point is OK.
For the second point, do you have an idea to support arbitrary function to calculate attention weight?

**Comment by kiyukuta** [comment]
> In my understanding, concat in (8) is
> Wa [ht; hs] = w1^T ht + w2^T hu
> but ht term cancels in softmax and has no sense.
> Is my understanding correct?

I think the `concat` (eq.8) in the Luong’s conference paper is wrong.
In their latest arXiv paper [1] and their implementation [2], the equation is as follows:
`Wa [ht; hs] =  va^T tanh( w1^T ht + w2^T hu )`.

This style (non-linear transformation) is also used in [Bahdanau+15]’s method which is an another famous paper[3]. 

[1] https://arxiv.org/pdf/1508.04025v5.pdf
[2] https://github.com/lmthang/nmt.hybrid/blob/master/code/layers/attnLayerForward.m#L44
[3] https://arxiv.org/pdf/1409.0473v7.pdf

**Comment by ywatanabex** [comment]
Hi @kiyukuta, thank you. 
The tanh expression does make sense.

**Comment by ywatanabex** [comment]
An easy way to support several types of attention scores is to split the attention function to two components.
In the attention function,
  Step 1.   compute scores using q and as.
  Step 2.   compute average of cs using the score

We can make implement three types of links for 1.
Also we can also make a function of 2.

The problem might be computational overhead.
In the step1. as is padded to matrix and compute scores and the score is changed to list of sequences.
Then in step2. the score is again padded to matrix form and the average is computed.
This conversion may take a little time.
Do you think this overhead is negligible? Or is there a good way of memory allocation?

**Comment by ywatanabex** [comment]
I split the attention function in two parts: `attention_score_dot` and `linear_combination`

```
y = attention(q, xs)
```

is the same as

```
 cs = attention_score_dot(q, xs)
 y = linear_combination(xs, cs)
```

**Comment by ywatanabex** [comment]
By the way, I'm not satisfied with the the function name. Do you have good idea?

File: chainer/functions/connection/attention.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+        gx = _c + _d + _e
+
+        gx_list = [gxi[:lens[i]] for (i, gxi) in enumerate(gx)]
+
+        return tuple([gq, ] + gx_list)
+
+
+def attention(q, xs):
+    """Attension from Vector to Sequence by inner products.
```
Comment (allanino) [inline chainer/functions/connection/attention.py]: Hi! There is a typo here: `Attension` should be `Attention`.

**Comment by MannyKayy** [comment]
Can we have the attention function dynamically consider variable length sequences across mini-batches (e.g. via an optional mask ) ?

**Comment by MannyKayy** [comment]
Bump, this would be a nice function to have in chainer v2.

**Pull request closed without merging**